gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/AckDeadlineRenewer.java

```diff
@@ -0,0 +1,268 @@
+      Long deadline = messageDeadlines.get(messageId);
+      if (deadline == null || deadline > threshold) {
+        // the message was removed (deadline == null) or removed and then added back
+        // (deadline > threshold), we should not renew its deadline (yet)
+        return new Message(messageId, null);
+      } else {
+        // Message deadline must be renewed, we must submit it again to the renewer
+        add(messageId.subscription(), messageId.ackId());
+        return new Message(messageId, deadline);
+      }
+    }
+  }
+
+  /**
+   * Adds a new message for which the acknowledge deadline should be automatically renewed. The
+   * message is identified by the subscription from which it was pulled and its acknowledge id.
+   * Auto-renewal will take place until the message is removed (see {@link #remove(String, String)}
+   * or {@link #remove(String, Iterable)}).
+   *
+   * @param subscription the subscription from which the message has been pulled
+   * @param ackId the message's acknowledge id
+   */
+  void add(String subscription, String ackId) {
+    synchronized (lock) {
+      long deadline = clock.millis() + MIN_DEADLINE_MILLISECONDS;
+      Message message = new Message(new MessageId(subscription, ackId), deadline);
+      messageQueue.add(message);
+      messageDeadlines.put(message.messageId(), deadline);
+    }
+  }
+
+  /**
+   * Adds new messages for which the acknowledge deadlined should be automatically renewed. The
+   * messages are identified by the subscription from which they were pulled and their
+   * acknowledge id. Auto-renewal will take place until the messages are removed (see
+   * {@link #remove(String, String)} or {@link #remove(String, Iterable)}).
+   *
+   * @param subscription the subscription from which the messages have been pulled
+   * @param ackIds the acknowledge ids of the messages
+   */
+  void add(String subscription, Iterable<String> ackIds) {
+    synchronized (lock) {
+      long deadline = clock.millis() + MIN_DEADLINE_MILLISECONDS;
+      for (String ackId : ackIds) {
+        Message message = new Message(new MessageId(subscription, ackId), deadline);
+        messageQueue.add(message);
+        messageDeadlines.put(message.messageId(), deadline);
+      }
+    }
+  }
+
+  /**
+   * Removes a message from this {@code AckDeadlineRenewer}. The message is identified by the
+   * subscription from which it was pulled and its acknowledge id. Once the message is removed from
+   * this {@code AckDeadlineRenewer}, automated ack deadline renewals will stop.
+   *
+   * @param subscription the subscription from which the message has been pulled
+   * @param ackId the message's acknowledge id
+   */
+  void remove(String subscription, String ackId) {
+    synchronized (lock) {
+      messageDeadlines.remove(new MessageId(subscription, ackId));
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
+    renewerFuture.cancel(false);
```

Comment:
Agree, I was also planning to change to `cancel(true)` and add a `closed` field to avoid doing the cleanup twice.